xc/device.ml

```diff
@@ -607,12 +607,34 @@ let qemu_media_change ~xs device _type params =
 
 	if is_upstream_qemu device.frontend.domid
 	then begin
+		let open Qmp in
 		let cd = "ide1-cd1" in
-		let qmp_cmd =
-			if params = ""
-			then (Qmp.Command(None, Qmp.Eject (cd, Some true)))
-			else (Qmp.Command(None, Qmp.Change (cd, params, None))) in
-		qmp_write device.frontend.domid qmp_cmd
+		match params with
+		| "" ->
+			let qmp_cmd = Command(None, Eject (cd, Some true)) in
+			qmp_write device.frontend.domid qmp_cmd
+		| _ ->
+			try
+				let c = Qmp_protocol.connect (Printf.sprintf "/var/run/xen/qmp-libxl-%d" device.frontend.domid) in
+				let fd_of_c = Qmp_protocol.to_fd c in
+				let fd_of_cd = Unix.openfile params [ Unix.O_RDONLY ] 0o640 in
+				ignore(Fd_send_recv.send_fd fd_of_c " " 0 1 [] fd_of_cd);
+				Unix.close fd_of_cd;
```

Comment:
This close will be missed if `Fd_send_recv.send_fd` raises an exception. It would be better to do this with a `finally` roughly like this:

```
let fd_of_cd = Unix.openfile params [ Unix.O_RDONLY ] 0o640 in
finally 
  (fun () -> Fd_send_recv.send_fd fd_of_c " " 0 1 [] fd_of_cd)
  (fun () -> Unix.close fd_of_cd);
```